flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergSourceSplitReader.java

```diff
@@ -80,6 +88,35 @@ public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
     }
 
     if (currentReader.hasNext()) {
+      if (pausedSplits.contains(currentSplitId)) {
+        // Wait until the reader is unblocked. Wake every second to catch any missed signal.
+        // Return empty records if wakeUp is called, so pauseOrResumeSplits could be processed.
+        boolean first = true;
+
+        while (pausedSplits.contains(currentSplitId)) {
+          if (first) {
+            LOG.info("Reader {} paused reading split {}", indexOfSubtask, currentSplitId);
+            first = false;
+          } else {
+            LOG.trace("Reader {} is still paused reading split {}", indexOfSubtask, currentSplitId);
+          }
+
+          synchronized (wakeUp) {
+            if (wakeUp.get()) {
+              wakeUp.set(false);
+              return new RecordsBySplits(Collections.emptyMap(), Collections.emptySet());
+            }
+
+            try {
+              wakeUp.wait(1000);
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+              throw new RuntimeException("Interrupted while paused due to watermark alignment.", e);
```

Comment:
@pvary sorry, I misspoken. I meant `IOException` instead of `InterruptedException`

Reply:
Understable. The `fetch` javadoc says: ` or it can just throw an interrupted exception`.
Set the type of the thrown exception to `IOException`